File: bindings/ruby/test/test_tree.rb
require 'minitest/autorun'
require_relative '../lib/cconfigspace'

class CConfigSpaceTestTree < Minitest::Test
  def test_create
    rng = CCS::Rng.new
    root = CCS::Tree.new(arity: 4, value: "foo")
    assert_equal( "foo", root.value )
    assert_equal( 4, root.arity)
    assert_nil( root.get_child(1) )
    root.children.each { |c|
      assert_nil( c )
    }
    assert_nil( root.parent )
    assert_nil( root.index )
    assert_equal( root.handle, root.get_node_at_position([]).handle )
    assert_equal( [], root.position )
    assert_equal( 0, root.depth )

    assert_equal( 1.0, root.weight )
    root.weight = 2.0
    assert_equal( 2.0, root.weight )
    assert_equal( 1.0, root.bias )
    root.bias = 2.0
    assert_equal( 2.0, root.bias )
    i = root.sample(rng)
    assert( i.nil? || (i >= 0 and i < 4) )
    root.samples(rng, 100).each { |i|
      assert( i.nil? || (i >= 0 and i < 4) )
    }

    child = CCS::Tree.new(arity: 3, value: "bar" )
    assert_equal( "bar", child.value )
    root.set_child(2, child)
    assert_equal( 2, child.index )
    assert_equal( ["foo", "bar"], child.values )
    assert_equal( [2], child.position )
    assert_equal( 1, child.depth )
    assert_equal( child.handle, root.get_child(2).handle )
    assert_equal( child.handle, root.get_node_at_position([2]).handle )
    assert_equal( ["foo", "bar"], root.get_values_at_position([2]) )

    buff = root.serialize
    tree = CCS.deserialize(buffer: buff)
    assert_equal( tree.object_type, :CCS_OBJECT_TYPE_TREE )
    assert_equal( ["foo", "bar"], tree.get_values_at_position([2]) )
  end
end

File: bindings/python/test/test_tree.py
import unittest
import re
import sys
sys.path.insert(1, '.')
sys.path.insert(1, '..')
import cconfigspace as ccs

class TestTree(unittest.TestCase):

  def test_create(self):
    rng = ccs.Rng()
    root = ccs.Tree(arity = 4, value = "foo")
    self.assertEqual( "foo", root.value )
    self.assertEqual( 4, root.arity)
    self.assertEqual( None, root.get_child(1) )
    for c in root.children:
      self.assertEqual( None, c )
    self.assertEqual( None, root.parent )
    self.assertEqual( None, root.index )
    self.assertEqual( root.handle.value, root.get_node_at_position([]).handle.value )
    self.assertEqual( [], root.position )
    self.assertEqual( 0, root.depth )

    self.assertEqual( 1.0, root.weight )
    root.weight = 2.0
    self.assertEqual( 2.0, root.weight )
    self.assertEqual( 1.0, root.bias )
    root.bias = 2.0
    self.assertEqual( 2.0, root.bias )
    i = root.sample(rng)
    self.assertTrue( i is None or (i >= 0 and i < 4) )
    for i in root.samples(rng, 100):
      self.assertTrue( i is None or (i >= 0 and i < 4) )

    child = ccs.Tree(arity = 3, value = "bar" )
    self.assertEqual( "bar", child.value )
    root.set_child(2, child)
    self.assertEqual( 2, child.index )
    self.assertEqual( ["foo", "bar"], child.values )
    self.assertEqual( [2], child.position )
    self.assertEqual( 1, child.depth )
    self.assertEqual( child.handle.value, root.get_child(2).handle.value )
    self.assertEqual( child.handle.value, root.get_node_at_position([2]).handle.value )
    self.assertEqual( ["foo", "bar"], root.get_values_at_position([2]) )

    buff = root.serialize()
    tree = ccs.Object.deserialize(buffer = buff)
    self.assertEqual( tree.object_type, ccs.ObjectType.TREE )
    self.assertEqual( ["foo", "bar"], tree.get_values_at_position([2]) )
